service/controller/v13/resource/instance/create_deployment_completed.go

```diff
@@ -0,0 +1,83 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/azure-operator/service/controller/v13/blobclient"
+	"github.com/giantswarm/azure-operator/service/controller/v13/key"
+	"github.com/giantswarm/azure-operator/service/controller/v13/resource/instance/internal/state"
+	"github.com/giantswarm/microerror"
+)
+
+func (r *Resource) deploymentCompletedTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+	deploymentsClient, err := r.getDeploymentsClient(ctx)
+	if err != nil {
+		return currentState, microerror.Mask(err)
```

Comment:
Why is the current state dispatched here? Is it supposed to be mutated inside the step implementations of the state machine? I think it is an anti pattern to return anything else than the zero value in case of an returned error. Nobody should ever expect anything being returned from a function call than the error, in case an error is returned.

Reply:
I leaned towards this approach, because `State` for now is just plain `string` and any string is valid state in general and I didn't want to create some special reserved words for error cases. Go also lacks nice way to express `Option` type and I wanted to comply in all possible cases with restriction that `TransitionFunc` must return configured state.

Reply:
The state machine is not doing anything with the state in error cases when I got that right. So we should simply return an empty string here and make clear that the state in error cases has no effects whatsoever.

Reply:
Ok. I changed that.